src/main/java/co/cask/hydrator/plugin/DynamicPartitionFileSetAvroSink.java

```diff
@@ -0,0 +1,164 @@
+
+  public DynamicPartitionFileSetAvroSink(DynamicPartitionAvroSinkConfig config) {
+    super(config);
+    this.config = config;
+  }
+
+  @Override
+  public void prepareRun(BatchSinkContext context) throws DatasetManagementException{
+    super.prepareRun(context);
+    Map<String, String> sinkArgs = getAdditionalPFSArguments();
+    PartitionedFileSetArguments.setDynamicPartitioner
+      (sinkArgs, DynamicPartitionFileSetAvroSink.FieldValueDynamicPartitioner.class);
+    context.addOutput(Output.ofDataset(config.name, sinkArgs));
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    super.initialize(context);
+    recordTransformer = new StructuredToAvroTransformer(config.schema, config.fieldNames);
+  }
+
+  @Override
+  public void transform(StructuredRecord input,
+                        Emitter<KeyValue<AvroKey<GenericRecord>, NullWritable>> emitter) throws Exception {
+    emitter.emit(new KeyValue<>(new AvroKey<>(recordTransformer.transform(input)), NullWritable.get()));
+  }
+
+  @Override
+  protected void addPartitionedFileSetProperties(PartitionedFileSetProperties.Builder properties) {
+    FileSetUtil.configureAvroFileSet(config.schema, properties);
+    properties.addAll(FileSetUtil.getAvroCompressionConfiguration(config.compressionCodec, config.schema, true));
+  }
+
+  @Override
+  protected DatasetProperties getDatasetProperties(Partitioning partitioning, Map.Entry<Schema, String> outputSchemaPair) {
+    PartitionedFileSetProperties.Builder properties = PartitionedFileSetProperties.builder();
+    addPartitionedFileSetProperties(properties);
+    return properties
+      .setPartitioning(partitioning)
+      .setBasePath(config.getNonNullBasePath())
+      .setTableProperty("avro.schema.literal",
+                        outputSchemaPair.getKey().toString())
+      .setExploreSchema(outputSchemaPair.getValue()
+                          .substring(1, outputSchemaPair.getValue().length() - 1))
+      .add(DatasetProperties.SCHEMA, config.schema)
+      .build();
+  }
+
+  /**
+   * Config for DynamicPartitionFileSetAvroSink
+   */
+  public static class DynamicPartitionAvroSinkConfig extends PartitionedFileSetSinkConfig {
+    public DynamicPartitionAvroSinkConfig(String name, String schema, String fieldNames,
+                                             @Nullable String basePath, @Nullable String compressionCodec) {
+      super(name, schema, fieldNames, basePath, compressionCodec);
+    }
+  }
+
+  /**
+   * Dynamic partitioner that creates partitions based on a list of fields in each record.
+   */
+  public static final class FieldValueDynamicPartitioner
+    extends DynamicPartitioner<AvroKey<GenericRecord>, NullWritable> {
+    private String[] fieldNames;
+
+    @Override
+    public void initialize(MapReduceTaskContext mapReduceTaskContext) {
+      if (mapReduceTaskContext.getPluginProperties(DynamicPartitionFileSetAvroSink.NAME) == null) {
```

Comment:
is this really needed? how was the pipeline published if the plugin doesn't exist?